core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,134 @@
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().capturedAs(MERGE_PROCESSOR_NODE_CAPTURE).with(source().matching(
+                                    project().with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN))).capturedAs(PROJECT_NODE_CAPTURE)))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+
+        ImmutableMap.Builder<String, Constant> assignmentsBuilder = ImmutableMap.builder();
+
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+
+            // we don't support yet any expressions in update statements, only constants
+            // also in case of UPDATE ALL statement, optimiser transforms expression to base64 call
+            if (assigmentNode instanceof Literal || assigmentNode instanceof SymbolReference || (assigmentNode instanceof Cast && ((Cast) assigmentNode).getExpression() instanceof Literal)) {
+                ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                        context.getSession(),
+                        ((Expression) assigmentNode),
+                        context.getSymbolAllocator().getTypes(),
+                        plannerContext,
+                        typeAnalyzer);
+
+                ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();
+
+                if (connectorExpression instanceof Constant constantExpression) {
```

Comment:
I think the code a few lines above : 
`if (assigmentNode instanceof Literal || assigmentNode instanceof SymbolReference || (assigmentNode instanceof Cast && ((Cast) assigmentNode).getExpression() instanceof Literal)) {`

add this check - as far as I see from debug : 
If Expression is Literal or Cast (with child Literal) - it will be translated to ConnectorExpression - Constant, which is our interested case.
If Expression is SymbolReference - it means that column is untouched, also valid case for us (just skipping this column).
All other cases are declined by this check (Like arithmetic expressions e.t.c)

Reply:
refactored and tests added